src/transformers/models/clvp/modeling_clvp.py

```diff
@@ -0,0 +1,1956 @@
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        outputs = self.model(
+            input_ids=input_ids,
+            past_key_values=past_key_values,
+            attention_mask=attention_mask,
+            token_type_ids=token_type_ids,
+            position_ids=position_ids,
+            head_mask=head_mask,
+            inputs_embeds=inputs_embeds,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        hidden_states = outputs[0]
+
+        lm_logits = self.final_norm(hidden_states)
+        lm_logits = self.lm_head(lm_logits)
+
+        loss = None
+        if labels is not None:
+            labels = labels.to(lm_logits.device)
+            # Shift so that tokens < n predict n
+            shift_logits = lm_logits[..., :-1, :].contiguous()
+            shift_labels = labels[..., 1:].contiguous()
+            # Flatten the tokens
+            loss_fct = CrossEntropyLoss()
+            loss = loss_fct(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
+
+        if not return_dict:
+            output = (lm_logits,) + outputs[1:]
+            return ((loss,) + output) if loss is not None else output
+
+        return CausalLMOutputWithCrossAttentions(
+            loss=loss,
+            logits=lm_logits,
+            past_key_values=outputs.past_key_values,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+            cross_attentions=outputs.cross_attentions,
+        )
+
+    @staticmethod
+    # Copied from transformers.models.gpt2.modeling_gpt2.GPT2LMHeadModel._reorder_cache
+    def _reorder_cache(
+        past_key_values: Tuple[Tuple[torch.Tensor]], beam_idx: torch.Tensor
+    ) -> Tuple[Tuple[torch.Tensor]]:
+        """
+        This function is used to re-order the `past_key_values` cache if [`~PreTrainedModel.beam_search`] or
+        [`~PreTrainedModel.beam_sample`] is called. This is required to match `past_key_values` with the correct
+        beam_idx at every generation step.
+        """
+        return tuple(
+            tuple(past_state.index_select(0, beam_idx.to(past_state.device)) for past_state in layer_past)
+            for layer_past in past_key_values
+        )
+
+
+@add_start_docstrings(
+    "The composite Clvp model with a text encoder, speech encoder and speech decoder model."
+    "The speech decoder model generates the speech_ids from the text and the text encoder and speech encoder works"
+    "together to filter out the best speech_ids.",
+    CLVP_START_DOCSTRING,
+)
+class ClvpModelForConditionalGeneration(ClvpPreTrainedModel):
```

Comment:
I don't think our class design is quite right here. IMO we should have the following classes defined:
1. ClvpConditionEncoder
2. ClvpEncoder
3. ClvpDecoder
4. ClvpModel: formed for ClvpConditionEncoder, ClvpEncoder, ClvpDecoder. Analogous to WhisperModel: https://github.com/huggingface/transformers/blob/1a2e966cfe916cdfd78ef7e1385911106bc33f50/src/transformers/models/whisper/modeling_whisper.py#L1227
5. ClvpForConditionalGeneration: formed of ClvpModel and an LM head on top. Analagous to WhisperForConditionalGeneration: https://github.com/huggingface/transformers/blob/1a2e966cfe916cdfd78ef7e1385911106bc33f50/src/transformers/models/whisper/modeling_whisper.py#L1395

=> this structure would bring the code in alignment with Whisper

Reply:
should I then remove the `ClvpForCausalLM`? Or keep it as a stack of `ClvpDecoder`s and a `LM head`.